src/AbstractOperations/AbstractOperations.jl

```diff
@@ -0,0 +1,95 @@
+module AbstractOperations
+
+export ∂x, ∂y, ∂z, @at, Computation, compute!, @unary, @binary, @polynary
+
+using Base: @propagate_inbounds
+
+using Oceananigans: @hascuda
+
+@hascuda using CUDAnative, CUDAdrv, CuArrays
+
+using Oceananigans, Adapt
+
+using Oceananigans: AbstractModel, AbstractGrid, AbstractField, AbstractLocatedField, Face, Cell, 
+                    xnode, ynode, znode, location, show_location, show_domain, short_show,
+                    device, launch_config, architecture, 
+                    HorizontalAverage, zero_halo_regions!, normalize_horizontal_sum!
+
+import Oceananigans: run_diagnostic, data, architecture
+
+using Oceananigans.TurbulenceClosures: ∂x_caa, ∂x_faa, ∂y_aca, ∂y_afa, ∂z_aac, ∂z_aaf, 
+                                       ▶x_caa, ▶x_faa, ▶y_aca, ▶y_afa, ▶z_aac, ▶z_aaf,
+                                       ▶xy_cca, ▶xy_ffa, ▶xy_cfa, ▶xy_fca, 
+                                       ▶xz_cac, ▶xz_faf, ▶xz_caf, ▶xz_fac, 
+                                       ▶yz_acc, ▶yz_aff, ▶yz_acf, ▶yz_afc,
+                                       ▶xyz_ccc, ▶xyz_fcc, ▶xyz_cfc, ▶xyz_ccf,
+                                       ▶xyz_fff, ▶xyz_ffc, ▶xyz_fcf, ▶xyz_cff
+
+using GPUifyLoops: @launch, @loop
+
+#####
+##### Basic functionality
+#####
+
+abstract type AbstractOperation{X, Y, Z, G} <: AbstractLocatedField{X, Y, Z, Nothing, G} end
+
+const ALF = AbstractLocatedField
+
+data(op::AbstractOperation) = op
+Base.parent(op::AbstractOperation) = op
```

Comment:
Though `AbstractField` does not subtype `AbstractArray`, it is often used as one. We use `parent` to obtain "underlying data" associated with a wrapper type. `AbstractOperation` is no different. Specifically, we want to be able to interchange fields, arrays, wrapper arrays, and abstract operations in `HorizontalAverage` (and probably other code in the future).

`data` is similar, but an internal implementation of the `Field` type. All `AbstractLocatedField` have an `A` type parameter that refers to the type of their "data". The function `data` returns that array when it exists, or it returns the object itself in the case of `AbstractOperation`, where there is no field `data`.

Reply:
I guess I should say that specifically, `data` is used when building an `AbstractOperation` (because unfortunately we still can't use `Field` on the GPU). An `AbstractOperation` can be performed on any `AbstractLocatedField`, which includes both `AbstractOperation`s and concrete `Field`s.